### DISCUSSION TITLE
It works ON Windows 11 but very slowly.

### DISCUSSION THREAD

**Discussion by ghost** [NiyaShy/XB1ControllerBatteryIndicator#90] [General]
It works very well, but it actually takes quite a while to recognize the device. Is there a way to make it connect faster? Also, I have a question: I'm in my first year as a developer and I'm learning to work with Java. My question is if this project could be done in Java language and be functional. Thank you very much for the program. Honestly, the Game Bar in Windows 11 works quite poorly. Best regards.

**Comment by NiyaShy** [comment]
Hey, since I still work with Win10 (and don't plan to upgrade to 11) I can't say if it's a general issue or "just you". Would be great if others could also give feedback.
Asking just in case: did you connect the controller via one of the many dongles that use Microsofts proprietary protocol or via Bluetooth? In the latter case it's an issue with Microsofts implementation and I can't do anything about it. see #49 for details about that.

Concerning Java: highly depends on if anyone managed to implement an XInput "bridge" to communicate with devices that use that protocol. Without that it would be hard to impossible to query the controller.

**Comment by Elinor10** [comment]
I use:
Windows 11 (version 23H2, latest updates)
Latest Xbox controller (special edition that was released 3 months ago, model 1914)
Official Microsoft dongle

My experience:
Indeed works on Windows 11, identifies the controller fast, but retrieving the battery data takes about 2-3 minutes (which is slow).

**Reply by NiyaShy** [reply]
Interesting. Is it always that slow or does it vary?
Reports that battery level detection sometimes takes longer have been a thing since I took over this project, and I'm not sure if I can do anything about it. The code simply asks the controller for the data, and if no response comes back it tries again after a few seconds.

**Reply by Elinor10** [reply]
I've just installed yesterday. From what I've seen so far - it's 2 minutes or more (one time took 5 minutes). 
When I open "Xbox Accessories" (the official Microsoft app for checking the battery level) it detects the battery level immediately while this tray app is still on the status "Controller One - Found but still waiting for battery data..." (I'm including a screenshot). 
Maybe it's possible to find what "Xbox Accessories" uses (for battery level detection) and use that?
Even if it stays slow I thank you for making this app. It's really useful. 

![Screenshot 2024-05-17 142251](https://github.com/NiyaShy/XB1ControllerBatteryIndicator/assets/102003456/0bf3c570-06df-4696-b613-dd040621696c)

**Reply by NiyaShy** [reply]
Microsofts own tools being faster has always been a thing as well. I don't know how they do it, and unless they make their stuff open-source and/or publish the source code there's no proper way of finding it out. But I guess they use some proprietary/undocumented API for it.

**Comment by neilser** [comment]
> When I open "Xbox Accessories" (the official Microsoft app for checking the battery level) it detects the battery level immediately while this tray app is still on the status "Controller One - Found but still waiting for battery data..." (I'm including a screenshot).

Just out of curiosity - did you try moving the sticks/triggers or pressing any buttons? I've found on W10 that I get quicker battery-level detection by doing that. (Having said that, it's clear that the official app is winning somehow.)

> Maybe it's possible to find what "Xbox Accessories" uses (for battery level detection) and use that?

I'd have thought so, yes, but @NiyaShy will know for certain (maybe that was how the original methods were developed).

**Reply by Elinor10** [reply]
I tried that but it seems to have no effect (on W11).

I got here when I was actually looking for a different feature - some sort of notification when charging is done (100% battery level) because the led indicator was omitted from the newest charging cables (so no way of knowing).

**Reply by NiyaShy** [reply]
> I'd have thought so, yes, but @NiyaShy will know for certain (maybe that was how the original methods were developed).

The "original methods" ( I guess you mean how the polling currently happens) uses the XInput API that Microsoft has published and documented. 

Oh, and about charge status: sadly not possible. The controller does not report that it's charging, the only way to tell would be the rising battery level. But since those states aren't very exact either ("full" doesn't mean 100% but is most likely a range in the 80-100% region) I could only guess when charging is done. Oh, and to mention it just in case: connecting the controller directly to your PC via USB doesn't work either because the controller then just reports "I'm wired" and IIRC even stops reporting a battery level.

**Reply by neilser** [reply]
> The "original methods" ( I guess you mean how the polling currently happens) uses the XInput API that Microsoft has published and documented.

Ahhh. I was imagining that you might have monitored the MS app to see how it did what it did.

On Linux I'd just use strace and ltrace, but on Windows I'm less certain that you can capture everything. However, Process Monitor from Sysinternals has been pretty effective at catching stuff for me in the past, and it appears that Logger is also worth a look (https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/logger-and-logviewer) though I've not tried that approach.

It seems crystal clear that the MS app is doing things that are not documented, so this might be the only way to work it out. (A low-effort and _very_ long-shot alternative would be to raise a support request and inform them that their published API can't do what their own app does, and ask for the docs for it ;-))
